apphub/NLP/neural_machine_translation/transformer.ipynb

```diff
@@ -0,0 +1,709 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "id": "potential-albert",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fastestimator.dataset.data import tednmt\n",
+    "\n",
+    "train_ds, eval_ds, test_ds = tednmt.load_data(data_dir, translate_option=\"pt_to_en\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "heavy-championship",
+   "metadata": {},
+   "source": [
+    "Now that the dataset is downloaded, let's check what the dataset looks like:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "northern-emperor",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "example source language:\n",
+      "entre todas as grandes privações com que nos debatemos hoje — pensamos em financeiras e económicas primeiro — aquela que mais me preocupa é a falta de diálogo político — a nossa capacidade de abordar conflitos modernos como eles são , de ir à raiz do que eles são e perceber os agentes-chave e lidar com eles .\n",
+      "\n",
+      "example target language:\n",
+      "amongst all the troubling deficits we struggle with today — we think of financial and economic primarily — the ones that concern me most is the deficit of political dialogue — our ability to address modern conflicts as they are , to go to the source of what they 're all about and to understand the key players and to deal with them .\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"example source language:\")\n",
+    "print(train_ds[0][\"source\"])\n",
+    "print(\"\")\n",
+    "print(\"example target language:\")\n",
+    "print(train_ds[0][\"target\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "connected-bleeding",
+   "metadata": {},
+   "source": [
+    "## Preprocessing the languages\n",
+    "\n",
+    "Since the text by itself cannot be recognized by computers, we need to perform a series of transformations to the text. Here are the steps:\n",
+    "1. Split the sentence into words or sub-words. For example, \"I love apple\" can be split into [\"I\", \"love\", \"apple\"]. Sometimes in order to represent more words, a word is further reduced into sub-words. For example, `tokenization` can be split into `token` and `_ization`. As a result, a word like \"civilization\" doesn't require extra space when both `civil` and `_ization` are already in the dictionary.\n",
+    "2. Map the tokens into a discrete index according to the dictionary. In this task, we are loading an pretrained tokenizer with a built-in dictionary already.\n",
+    "3. Add a [start] and [end] index around every token indexes, this is mainly to help the network understand when the beginning and ending of a sentence.\n",
+    "4. When creating a batch of multiple sentences, pad the data according to the maximum length within the batch with 0."
```

Comment:
```suggestion
    "4. When creating a batch of multiple sentences, pad the shorter sentences with 0 so that each sentence in the batch has the same length."
```